### ISSUE TITLE
Add tox

### ISSUE THREAD

**Issue by adriangonz** [SeldonIO/seldon-core#1042]
## Context

When we run the tests on the Python package, we always run them on top of the checked-out source. Therefore, the environment won't be the same as when the `seldon_core` package is installed from `pip`. This means that some issues may slip through as was the case on #1040. 

A way to avoid this is to run the tests on fresh virtual environments where the package is obtained uniquely from what's declared in `setup.py`.

## Suggested solution

[`tox`](https://tox.readthedocs.io) is a tool that allows you to run your tests across different environments. By default, on each of these environments the source will be built from the output of `python setup.py sdist`. This way the testing environment matches what end users will have when they install `seldon_core` from `pip`.

On top of this, it also allows you to define multiple testing environments. Therefore, we could move the testing of `{ seldon_core, seldon_core[tensorflow], seldon_core[gcs], seldon_core[all] }` from the `jenkins-x.yaml` file (where it currently is) to the `python/setup.cfg` file.

The caveats are:

- We would be introducing yet another new tool to the workflow.
- Since `tox` runs the tests across multiple environments, it is way slower tan just running `pytest`. This can be solved by using `pytest` during the regular development workflow and only running `tox` as part of the CI pipeline.

**Comment by adriangonz** [comment]
@axsaucedo @cliveseldon @jklaise what are your thoughts on introducing `tox` (on top of `pytest`) to run `seldon_core`'s tests?

**Comment by jklaise** [comment]
@adriangonz I looked into it briefly some time ago, I think it was tricky to run tox together with travis across multiple different python versions, but it's worth re-visiting. For alibi we specify different Python versions straight into travis, so no special need for tox. Although I guess we still have the issue that tests are run on master not on pip-installable versions (but before a release this would've been covered by the latest tests on master?)

**Comment by adriangonz** [comment]
@jklaise there are some edge-cases that wouldn't be covered by running the tests on `master`. An example is issue #1040, where I forgot to add a couple new files to the `MANIFEST.in` which weren't getting picked by `setup.py`. When we ran the tests on the checked-out source, since those files were present there everything looked fine. However, if we had run the tests with the files installed through `pip`, that would have got caught.

For `seldon_core` I think we don't worry much about different Python versions. I actually still don't know where `tox` gets the Python versions from. I know that `tox` allows you to configure an `envlist` with things like `py37`, `py36`, etc. but I've got no idea where it gets them from at test time. 

Instead, the main concerns that this issue tries to address is that the tests run on `pip`-ed source and that we test the different combinations of the extra dependencies (which are currently `tensorflow`, `gcs` and `all`).
